Gems/AtomLyIntegration/CommonFeatures/Code/Source/PostProcess/EditorPostFxSystemComponent.cpp

```diff
@@ -95,18 +95,33 @@ namespace AZ
         {
             AZ_UNUSED(catalogFile);
 
-            // automatically register all layer categories assets
-            AZ::Data::AssetCatalogRequestBus::Broadcast(&AZ::Data::AssetCatalogRequestBus::Events::EnumerateAssets,
-                nullptr,
-                [this](const AZ::Data::AssetId assetId, const AZ::Data::AssetInfo& assetInfo) {
+            // automatically register all layer category assets at Editor startup
+
+            AZStd::vector<AZ::Data::AssetId> layerAssetIds;
+
+            // First run through all the assets and gather up the asset IDs for all layer assets
+            AZ::Data::AssetCatalogRequestBus::Broadcast(
+                &AZ::Data::AssetCatalogRequestBus::Events::EnumerateAssets, nullptr,
+                [&layerAssetIds](const AZ::Data::AssetId assetId, const AZ::Data::AssetInfo& assetInfo)
+                {
                     const auto assetType = azrtti_typeid<EditorPostFxLayerCategoriesAsset>();
                     if (assetInfo.m_assetType == assetType)
                     {
-                        m_layerCategoriesAssetMap[assetId] = AZ::Data::AssetManager::Instance().GetAsset(assetId, assetType, AZ::Data::AssetLoadBehavior::PreLoad);
-                        m_layerCategoriesAssetMap[assetId].QueueLoad();
+                        layerAssetIds.emplace_back(assetId);
```

Comment:
The invocation of the callback should probably inside of AssetCatalogRequestBus EnumerateAssets should probably done outside of the `DispatchLockGuard` mutex to prevent deadlocking from occurring.

Alternatively if we created an AZ::Event for that mirrored each the AssetCatalogRequestBus event functions, we could iteratively transition the existing call sites over to use those functions.

I think we could get away with unlock the DispatchMutex inside of [EnumerateAssets](https://github.com/o3de/o3de/blob/development/Code/Framework/AzFramework/AzFramework/Asset/AssetCatalog.cpp#L487-L511) since it uses another mutex to protect the AssetRegistry inside of it.

Here is snippet of the change to AssetCatalog::EnumerateAssets
```c++
//=========================================================================
// EnumerateAssets
//=========================================================================
void AssetCatalog::EnumerateAssets(BeginAssetEnumerationCB beginCB, AssetEnumerationCB enumerateCB, EndAssetEnumerationCB endCB)
{
    using AssetCatalogRequestBusContext = typename AZ::Data::AssetCatalogRequestBus::Context;
    using AssetCatalogRequestBusDispatchScopedLock = typename AZ::Data::AssetCatalogRequestBus::Context::DispatchLockGuard;
    // Setting  trackCallstack to true causes the context mutex to attempt to re-lock
    // That is being avoided here as the code only wants to unlock the Context mutex if it is in a dispatch
    constexpr bool trackCallstack = false;
    AssetCatalogRequestBusContext* assetCatalogContext
        = AZ::Data::AssetCatalogRequestBus::GetContext(trackCallstack);

    bool hasAssetCatalogMutex = false;
    if (assetCatalogContext != nullptr &&
        AZ::Data::AssetCatalogRequestBus::IsInDispatchThisThread(assetCatalogContext))
    {
        hasAssetCatalogMutex = true;
        // Unlock the dispatch mutex for the AssetCatalogRequestBus
        assetCatalogContext->m_contextMutex.unlock();
    }
    if (beginCB)
    {
        beginCB();
    }

    if (enumerateCB)
    {
        AZStd::lock_guard<AZStd::recursive_mutex> lock(m_registryMutex);

        for (auto& it : m_registry->m_assetIdToInfo)
        {
            enumerateCB(it.first, it.second);
        }
    }

    if (endCB)
    {
        endCB();
    }

    if (hasAssetCatalogMutex)
    {
        // Relock the mutex if it was unlocked earlier
        assetCatalogContext->m_contextMutex.lock();
    }
}
```